docs/feature_autocorrect.md

```diff
@@ -0,0 +1,281 @@
+    switch (*keycode) {
+        // Exclude these keycodes from processing.
+        case KC_LSFT:
+        case KC_RSFT:
+        case KC_CAPS:
+        case QK_TO ... QK_ONE_SHOT_LAYER_MAX:
+        case QK_LAYER_TAP_TOGGLE ... QK_LAYER_MOD_MAX:
+        case QK_ONE_SHOT_MOD ... QK_ONE_SHOT_MOD_MAX:
+            return false;
+
+        // Mask for base keycode from shifted keys.
+        case QK_LSFT ... QK_LSFT + 255:
+        case QK_RSFT ... QK_RSFT + 255:
+            if (*keycode >= QK_LSFT && *keycode <= (QK_LSFT + 255)) {
+                *mods |= MOD_LSFT;
+            } else {
+                *mods |= MOD_RSFT;
+            }
+            *keycode &= 0xFF; // Get the basic keycode.
+            return true;
+#ifndef NO_ACTION_TAPPING
+        // Exclude tap-hold keys when they are held down
+        // and mask for base keycode when they are tapped.
+        case QK_LAYER_TAP ... QK_LAYER_TAP_MAX:
+#    ifdef NO_ACTION_LAYER
+            // Exclude Layer Tap, if layers are disabled
+            // but action tapping is still enabled.
+            return false;
+#    endif
+        case QK_MOD_TAP ... QK_MOD_TAP_MAX:
+            // Exclude hold if mods other than Shift is not active
+            if (!record->tap.count) {
+                return false;
+            }
+            *keycode &= 0xFF;
+            break;
+#else
+        case QK_MOD_TAP ... QK_MOD_TAP_MAX:
+        case QK_LAYER_TAP ... QK_LAYER_TAP_MAX:
+            // Exclude if disabled
+            return false;
+#endif
+        // Exclude swap hands keys when they are held down
+        // and mask for base keycode when they are tapped.
+        case QK_SWAP_HANDS ... QK_SWAP_HANDS_MAX:
+#ifdef SWAP_HANDS_ENABLE
+            if (*keycode >= 0x56F0 || !record->tap.count) {
+                return false;
+            }
+            *keycode &= 0xFF;
+            break;
+#else
+            // Exclude if disabled
+            return false;
+#endif
+        // Handle custom keycodes
+        case QMKBEST:
+            return false;
+        case QMKLAYER:
+            *typo_buffer_size = 0;
+            return false;
+    }
+
+    // Disable autocorrect while a mod other than shift is active.
+    if ((*mods & ~MOD_MASK_SHIFT) != 0) {
+        *typo_buffer_size = 0;
+        return false;
+    }
+
+    return true;
+}
+```
+
+
+?> In this callback function, `return false` will skip processing of that keycode for autocorrect. Adding `*typo_buffer_size = 0` will also reset the autocorrect buffer at the same time, cancelling any current letters already stored in the buffer.
+
+### Apply Autocorrect
+
+Additionally, `apply_autocorrect(uint8_t backspaces, const char *str)` allows for users to add additional handling to the autocorrection, or replace the functionality entirely. This passes on the number of backspaces needed to replace the words, as well as the replacement string (partial word, not the full word).
```

Comment:
Nice! The possibility to play a sound when autocorrect triggers is really cool.

I suggest to work into this paragraph that the `str` arg is a pointer to PROGMEM data, so the callback needs to use `send_string_P()`, not `send_string()`.

string pointer

Reply:
I had a couple of motivations for this.  Such as somebody mentioning counting the number of corrections occurring, feedback when errors occur, etc.  

Also, I wanted a way to hook into my unicode typing stuff.  :)
𝓫𝓮𝓬𝓪𝓾𝓼𝓮 𝓽𝓱𝓲𝓼 𝓼𝓽𝓾𝓯𝓯 𝓲𝓼 𝓯𝓾𝓷.

but good point about the str parameter.

Reply:
Do you think the example can also be improved with a `return true` use case? Display correction on OLED but let the engine handle the correction?

Reply:
yeah, it could be worth it to have both.

Reply:
> 𝓫𝓮𝓬𝓪𝓾𝓼𝓮 𝓽𝓱𝓲𝓼 𝓼𝓽𝓾𝓯𝓯 𝓲𝓼 𝓯𝓾𝓷.

I love your unicode typing! Great motivation.